[datafusion] Connection pooling

CheyuWu opened the pull request:
<!-- .github/PULL_REQUEST_TEMPLATE.md -->

## Summary

Briefly explain what this PR does.
- Create a connection pool in `src/moonlink_datafusion/src/connection_pool.rs`
- Add a test to test the connection pool


## Related Issues

Closes #1374 

## Changes

- Create `src/moonlink_datafusion/src/connection_pool.rs`
- Use connection pool in `src/moonlink_datafusion/src/table_provider.rs`

## Checklist

- [ ] Code builds correctly
- [x] Tests have been added or updated
- [ ] Documentation updated if necessary
- [x] I have reviewed my own changes

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce a global UnixStream connection pool and use it in the table provider; add tests and wiring.
> 
> - **Core**:
>   - Implement global connection pool in `src/moonlink_datafusion/src/connection_pool.rs` with `Pool` and `PooledStream` (idle timeout, max entries per-URI, reuse on drop).
> - **Integration**:
>   - Replace direct `UnixStream` usage in `src/moonlink_datafusion/src/table_provider.rs` with pooled connections (`Pool::get_stream`, pass `PooledStream`, update drop to return to pool).
>   - Expose module in `src/moonlink_datafusion/src/lib.rs`.
> - **Tests**:
>   - Add async tests for basic reuse, multi-URI concurrency, and max-capacity behavior in `connection_pool.rs`.
> - **Deps**:
>   - Add `tempfile` as `dev-dependencies` in `src/moonlink_datafusion/Cargo.toml`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e0c543ac7c64d8aebc06c7c294364b135af212d1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

dpxcc commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,44 @@
+static CONN_POOL: LazyLock<Mutex<Vec<UnixStream>>> = LazyLock::new(|| Mutex::new(Vec::new()));
+
+pub(crate) async fn get_stream(uri: &str) -> std::io::Result<UnixStream> {
+    if let Some(stream) = CONN_POOL.lock().unwrap().pop() {
```
You are putting connections of different uris in the same pool, so there is no guarantee that you will get a correct connection
We should make the connection pool scope to catalog, which has a fixed uri

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,44 @@
+
+static CONN_POOL: LazyLock<Mutex<Vec<UnixStream>>> = LazyLock::new(|| Mutex::new(Vec::new()));
+
+pub(crate) async fn get_stream(uri: &str) -> std::io::Result<UnixStream> {
```
Should we use `crate::Result` here? otherwise we loss error source location

dpxcc commented on src/moonlink_datafusion/src/table_provider.rs:
```diff
@@ -39,7 +40,7 @@ pub struct MooncakeTableProvider {
 impl MooncakeTableProvider {
     pub async fn try_new(uri: &str, schema: String, table: String, lsn: u64) -> Result<Self> {
-        let mut stream = UnixStream::connect(uri).await?;
+        let mut stream = get_stream(uri).await?;
```
This works. But for a query joining many tables, each table will get one connection. I think we can improve this so that we use one connection per query per catalog

dpxcc commented on src/moonlink_datafusion/Cargo.toml:
```diff
@@ -18,5 +18,6 @@ moonlink_table_metadata = { path = "../moonlink_table_metadata" }
 object_store = { workspace = true }
 parquet = { workspace = true }
 roaring = { workspace = true }
+tempfile = { workspace = true }
```
This is test only

dpxcc commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,44 @@
+use std::sync::{LazyLock, Mutex};
+
```
nit: remove new line

dpxcc commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,44 @@
+
+use tokio::net::UnixStream;
+
+static CONN_POOL: LazyLock<Mutex<Vec<UnixStream>>> = LazyLock::new(|| Mutex::new(Vec::new()));
```
nit: let's just name it POOL

dpxcc commented on src/moonlink_datafusion/src/table_provider.rs:
```diff
@@ -202,7 +203,7 @@ impl ParquetFileReaderFactory for MooncakeParquetFileReaderFactory {
 
 #[derive(Debug)]
-struct MooncakeTableScan {
+pub(crate) struct MooncakeTableScan {
```
no need to `pub(crate)`

CheyuWu commented:
Hi @dentiny @dpxcc 
- I have moved [tempfile](https://github.com/Mooncake-Labs/moonlink/pull/1801/files#diff-2b71c503e045161b9f985586a701c893abdd08f617674d56475f9da522565ae7R23-R25) to dev-dependency
- Address the nit problem
- Create a [struct](https://github.com/Mooncake-Labs/moonlink/pull/1801/files#diff-cc4ac496ce89ed13b4dbf693e301b4178be09673c87291d08420147606a0940fR5-R14) to store uri
- Use HashMap to address the different URIs in [here](https://github.com/Mooncake-Labs/moonlink/pull/1801/files#diff-cc4ac496ce89ed13b4dbf693e301b4178be09673c87291d08420147606a0940fR16-R17)
- Add concurrent stream test in [here](https://github.com/Mooncake-Labs/moonlink/pull/1801/files#diff-cc4ac496ce89ed13b4dbf693e301b4178be09673c87291d08420147606a0940fR68-R119)

PTAL

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,120 @@
+            return Ok(PooledStream::new(uri.to_string(), stream));
+        }
+    }
+    let stream = UnixStream::connect(uri).await?;
```
Please don't put IO operation inside of critical section

dentiny replied:
Some other considerations on connection pool:
- Do you want to set timeout to connection entries? If you do, do we need background maintainance thread?
- Do you want to set a max cap for connection pool?

CheyuWu replied:
OK, no problem

CheyuWu replied:
Thanks for the suggestions!
- For timeout, I plan to add a timeout directly to `UnixStream::connect` using `tokio::time::timeout`.
- For the connection pool, I will set a maximum cap per URI (10 -50  pooled connections per URI)

dentiny replied:
no pressure :)

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,120 @@
+static POOL: LazyLock<Mutex<HashMap<String, Mutex<Vec<UnixStream>>>>> =
+    LazyLock::new(|| Mutex::new(HashMap::new()));
+
+pub(crate) async fn get_stream(uri: &str) -> crate::Result<PooledStream> {
```
Do you think it possible to provide a RAII wrapper, which gets or create a connection on construction, and returns back on destruction? That way we never forget return.

dentiny commented on src/moonlink_datafusion/src/table_provider.rs:
```diff
@@ -39,10 +39,11 @@ pub struct MooncakeTableProvider {
     pub async fn try_new(uri: &str, schema: String, table: String, lsn: u64) -> Result<Self> {
-        let mut stream = UnixStream::connect(uri).await?;
-        let table_schema = get_table_schema(&mut stream, schema.clone(), table.clone()).await?;
+        let mut pooled_stream = get_stream(uri).await?;
```
If the returned cached connection is broken, how do we handle retry?

CheyuWu replied:
We can wrap the pooled connection in a RAII-style struct so it is automatically returned to the pool when dropped.

When getting a stream or calling `get_table_schema`, if the connection is broken, we retry a few times with backoff (e.g., fixed or exponential delay) before giving up.

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,120 @@
+
+#[derive(Debug)]
+pub struct PooledStream {
+    pub uri: String,
```
Not sure why do we need uri here?

dentiny commented on src/moonlink_datafusion/src/table_provider.rs:
```diff
@@ -230,14 +237,15 @@ impl MooncakeTableScan {
-            let mut stream = stream.expect("stream should be set by try_new");
-            if let Err(e) = scan_table_end(&mut stream, schema, table).await {
+            let mut pooled_stream = pooled_stream.expect("stream should be set by try_new");
+            if let Err(e) = scan_table_end(&mut pooled_stream.stream, schema, table).await {
```
If the rpc invocation fails, do we want to return connection back to pool still?

CheyuWu replied:
I think we can just drop it

CheyuWu replied:
I think we have two reasonable options:
- Directly drop the connection if the RPC fails, assuming it might be broken.

- Optionally perform a lightweight “ping” or health check to verify whether the connection is still healthy before returning it to the pool.

Which approach do you prefer?

CheyuWu commented:
Hi @dentiny PTAL
In the latest PR, I have
- Introduced a RAII wrapper for PooledStream, so we no longer need to explicitly call return_stream at the end of the lifecycle.
- Moved I/O operations out of critical sections to avoid blocking the pool lock.
- Added a background maintenance thread to handle idle timeouts.
- Enforced a maximum capacity for the connection pool.
- Improved connection health check in get_stream by removing broken connections.
- Added a retry mechanism (up to two attempts) in MooncakeTableProvider::try_new to increase robustness.

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+    last_used: Instant,
+}
+
+static POOL: LazyLock<Mutex<HashMap<String, Mutex<Vec<PooledEntry>>>>> =
```
nit: add a documentation, key is URI, value is a pool of entries.

dentiny replied:
we already have a global lock for `POOL`, curious what's the purpose for mutex inside of each entry?

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+
+const MAX_PER_URI: usize = 30;
+
+static IDLE_TIMEOUT_MS: AtomicU64 = AtomicU64::new(30_000);
```
I only see it updated in testing, which we should inject via connection pool configs, instead of global variables.

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+                let mut pool = POOL.lock().await;
+                let mut pool_vec = pool.entry(uri).or_default().lock().await;
+
+                if pool_vec.len() < MAX_PER_URI {
```
so every single time you push to the end and pop from the end of the vector;
on insufficient connection stream space, should we pop from the front? since it's least recently accessed, otherwise it's easy to get them timed-out anyway

CheyuWu replied:
Great idea, I like that, thx

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+}
+
+#[derive(Debug)]
+pub struct PooledStream {
```
curious why do you call it `PooledStream`?

CheyuWu replied:
Well, this is named by ChatGPT. I am poor at naming the function or variable

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+#[derive(Debug)]
+pub struct PooledStream {
+    pub uri: String,
+    pub stream: Option<UnixStream>,
```
nit: add a comment that option is made for ownership transfer at drop

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+
+struct PooledEntry {
+    stream: UnixStream,
+    last_used: Instant,
```
Incorrect variable naming I think, it's the timestamp where the entry is inserted into the pool, not the timestamp it was last used.

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+            vec.retain(|entry| entry.last_used.elapsed() <= idle_timeout());
+
+            while let Some(entry) = vec.pop() {
+                if is_connection_alive(&entry.stream).await {
```
please don't put IO operations inside of critical section
for implementation easiness, we could return the stream directly

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+        if let Some(mutex_vec) = pool.get_mut(uri) {
+            let mut vec = mutex_vec.lock().await;
+            // Remove expired streams
+            vec.retain(|entry| entry.last_used.elapsed() <= idle_timeout());
```
let's use deque for perf

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+        sleep(Duration::from_secs(5)).await;
+
+        let mut pool = POOL.lock().await;
+        for (_uri, vec_mutex) in pool.iter_mut() {
```
nit
```suggestion
        for (_, vec_mutex) in pool.iter_mut() {
```

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,361 @@
+    }
+}
+
+pub async fn start_maintenance_task() {
```
I'm fine with a background maintenance task, but can we scope it to pool?

CheyuWu commented:
Hi @dentiny, I have addressed the problem that you have mentioned, PTAL.

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,276 @@
+const DEFAULT_MAX_ENTRIES_PER_URI: usize = 30;
+const DEFAULT_IDLE_TIMEOUT_MS: u64 = 30_000;
+
+pub(crate) static POOL: LazyLock<Arc<Pool>> = LazyLock::new(|| {
```
All connection pool related variables, classes and functions seems to be scoped in the file, which means they could be `private`?

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,276 @@
+
+        if let Some(vec) = pool_inner.get_mut(uri) {
+            // Remove expired streams
+            vec.retain(|entry| entry.inserted_at.elapsed() <= pool.idle_timeout());
```
The entry is sorted in insertion time, so we don't need to iterate through all items.
could do sth like
```rs
// get the first item which is larger than timeout cutoff
// remove all items before
```

CheyuWu replied:
Should we change Vecque and use binary search to remove the timeout stream? Or is it better to use linear to delete the timeout stream?

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,276 @@
+    }
+}
+
+pub(crate) async fn get_stream_with_pool(
```
(no request to change, just a question)
Do you think it better to make it a member function for `Pool`?

CheyuWu replied:
You are right !! 
Making it a member function of `Pool` would make the API more idiomatic and easier to use.
Should I refactor it into a method in this PR or in the following PR?

CheyuWu commented:
Hi @dentiny, I have addressed the problem PTAL
- Make `DEFAULT_IDLE_TIMEOUT_MS` a duration constant
- All connection pool related variables, classes and functions seems to be scoped in the file
- optimize the removal of  `idle_timeout` stream
> Use linear to delete the timeout stream

CheyuWu commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,280 @@
+
+                pool_vec.push_back(PooledEntry {
+                    stream,
+                    inserted_at: Instant::now(),
```
Maybe we can have a follow-up PR to introduce a background thread that maintains the timeout?

dentiny commented on src/moonlink_datafusion/src/connection_pool.rs:
```diff
@@ -0,0 +1,280 @@
+
+        // Retrieve streams for both URIs again; should reuse connections from the pool
+        let (stream1b, stream2b) = tokio::join!(
+            test_pool.get_stream_with_pool(uri1),
```
> If we intentionally use an invalid path, the first connect fails, and we cannot obtain an initial stream to later verify reuse.

I only asked to use invalid URI here, not everywhere.

CheyuWu commented:
> @CheyuWu ping me if you want me to pick it up, thanks.
Hi @dentiny Thanks for offering! I’ll take this one, but I’ll reach out if I need any help.
